Manage dependencies using jsonnet-bundler

corvus-ch opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and
review the checklist below.

Contributors guide: ./CONTRIBUTING.md
-->

Resolves #77 

## Checklist
<!--
Remove items that do not apply. For completed items, change [ ] to [x].
-->

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Update the ./CHANGELOG.md.
- [x] Link this PR to related issues.

<!--
NOTE: these things are not required to open a PR and can be done afterwards,
while the PR is open.
-->

srueg reviewed (changes requested):
Now the component compilation fails because `dependencies/` is not in the search path anymore...

tobru commented on README.md:
````diff
@@ -64,6 +65,9 @@ using `helm template`.
       ```
 
+    Install jsonnet-bundler according to upstream [documentation](https://github.com/jsonnet-bundler/jsonnet-bundler#install).
+    For OS X, jsonnet-bundler is available as [Homebrew](https://brew.sh/) package.
````
I'd prefer to leave this line out, otherwise it gives the feeling that it's OS X biased or we would need to add the info for other OS'es as well. I'd leave how to install it up to the user.

tobru commented on docs/modules/ROOT/pages/index.adoc:
```diff
@@ -18,6 +18,14 @@ templates. Commodore sets up an https://kapitan.dev/inventory/[inventory] by
 downloading all required components and symlinking the relevant classes into
 the inventory structure which Kapitan understands.
 
+Components can define dependencies to Jsonnet libraries using https://github.com/jsonnet-bundler/jsonnet-bundler[jsonnet-bundler].
```
This feels a bit scarce. Shouldn't we add a bit more documentation (example, how-to, ...) about this feature, as it feels to be an important feature for the project.

tobru reviewed (changes requested):
Other than the inline comments: Should Jsonnet Bundlet not added to the Dockerfile as well?

simu commented on commodore/dependency_mgmt.py:
```diff
@@ -195,3 +197,39 @@ def fetch_jsonnet_libs(config, libs):
+    with open("jsonnetfile.json", "w") as file:
+        file.write(json.dumps(jsonnetfile, indent=4))
+
+    if call(['jb', 'install']) != 0:
```
Probably want to produce a useful error here if `jb` cannot be found

simu reviewed (changes requested):
Please fix the test failures, looks like component compile breaks with these changes

corvus-ch commented:
> Now the component compilation fails because dependencies/ is not in the search path anymore...
> We either add it to the search path for component compilation only or run jb also there.

I fixed this by splitting the creation of the `jsonnetfile.json` and the call to `jb`. I also removed `vendor` from the default search path and only add it where needed.

For the `commodore component compile`, `jb` is called for only that component. Search paths are adapted accordingly.

srueg commented on Dockerfile:
```diff
@@ -41,6 +41,10 @@ COPY --from=builder /usr/local/lib/python3.8/site-packages/kapitan ./kapitan
 
+FROM golang:1.15 AS jsonnet-bundler
+
+RUN go get github.com/jsonnet-bundler/jsonnet-bundler/cmd/jb
```
Let's download a specific version from the [releases](https://github.com/jsonnet-bundler/jsonnet-bundler/releases).

srueg commented on commodore/component/compile.py:
```diff
@@ -90,6 +92,8 @@ def compile_component(config: Config, component_path, value_files, search_paths,
 
         # Fetch Jsonnet libs
         fetch_jsonnet_libs(config, libs)
+        if os.path.exists(os.path.join(component_path, 'jsonnetfile.json')):
```
Since `component_path` is already a path type, it should be possible to do
```suggestion
        if (component_path / 'jsonnetfile.json').exists():
```

simu replied:
looks like the suggestion was not fully applied

corvus-ch replied:
```suggestion
        if (component_path / 'jsonnetfile.json').exists():
```

srueg commented on commodore/component/compile.py:
```diff
@@ -25,6 +25,8 @@ def compile_component(config: Config, component_path, value_files, search_paths,
     value_files = [P(f).resolve() for f in value_files]
     search_paths = [P(d).resolve() for d in search_paths]
+    search_paths.append('./dependencies/')
+    search_paths.append(os.path.join(component_path, 'vendor'))
```
```suggestion
    search_paths.append(component_path / 'vendor')
```

simu replied:
looks like the suggestion was not fully applied

simu commented on Dockerfile:
```diff
@@ -44,6 +44,7 @@ RUN ./kapitan/inputs/helm/build.sh \
 FROM base AS runtime
 
 RUN apt-get update && apt-get install -y --no-install-recommends \
+      curl \
```
That's gonna improve the debugging experience when running Commodore in Docker quite a bit :+1:

simu commented on docs/modules/ROOT/pages/dependencies.adoc:
```diff
@@ -0,0 +1,46 @@
+Those dependencies are resolved and downloaded using https://github.com/jsonnet-bundler/jsonnet-bundler[jsonnet-bundler].
+
+Define dependencies in a `jsonnetfile.json` file.
+This file must be placed at the root of the components file structure.
```
```suggestion
This file must be placed at the root of the component's file structure.
```

simu commented on docs/modules/ROOT/pages/dependencies.adoc:
```diff
@@ -0,0 +1,46 @@
+[TIP]
+====
+When compiling components with `commodore component compile`, jsonnet-bundler will create the file `jsonnetfile.lock.json` and the folder `vendor`.
+Add them to your `.gitignore`.
```
Probably worth updating the `.gitignore` in the component template? Doesn't hurt if the entries are there but not used.

simu reviewed (approved):
LGTM, apart from some not yet applied suggestions from the last round of code review

corvus-ch commented:
>  […] apart from some not yet applied suggestions from the last round of code review

With greetings from force push. I like that GitHub reopened those.

srueg commented on docs/modules/ROOT/pages/dependencies.adoc:
```diff
@@ -0,0 +1,46 @@
+====
+Jsonnet-bundler uses the file `jsonnetfile.lock.json` to freeze dependency versions for a project.
+Lock files of libraries aren't taken into consideration and within this setup, components are libraries.
+If your component require a specific version, you have to declare this in the `version` field within `jsonnetfile.json`.
```
```suggestion
If your component requires a specific version, you have to declare this in the `version` field within `jsonnetfile.json`.
```

simu commented:
> > […] apart from some not yet applied suggestions from the last round of code review
> 
> With greetings from force push. I like that GitHub reopened those.

I actually reopened them manually rather than write a new identical suggestion. GitHub isn't quite that smart yet unfortunately.
